### PR TITLE
added warning for distributedsampler in case of evaluation

### LINKED ISSUE #11364: Add a warning regarding distributed sampling if `trainer.test/validate` is called under distributed env.
Neat ! Let's make this a warning.

_Originally posted by @tchaton in https://github.com/PyTorchLightning/pytorch-lightning/pull/11198#discussion_r778899246_

### Motivation

Add a warning to let users know if they are running under distributed env and DistributedSampler is added within lightning in case of test/validate it should warn users that DistributedSampler adds extra samples to make sure all devices get equivalent batch sizes and accumulated metrics might not be correct and suggest them to use single device strategy.

### Additional context

Ques:
Should we warn in case user adds the DistributedSampler manually with `Trainer(replace_sampler_ddp=False)`?

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @awaelchli @rohitgr7 @justusschock @akihironitta @ninginthecloud

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #11364

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/common/test_set.rst [hunk 1]
```diff
@@ -144,5 +144,12 @@ Apart from this ``.validate`` has same API as ``.test``, but would rely respecti
     ``.validate`` method uses the same validation logic being used under validation happening within
     :meth:`~pytorch_lightning.trainer.trainer.Trainer.fit` call.
 
+.. warning::
+
+    When using `trainer.validate()`, it is recommended to use single device since Distributed Training such as DDP internally
```
Comment (rohitgr7) [inline docs/source/common/test_set.rst:149]:
```suggestion
    When using ``trainer.validate()``, it is recommended to use a single device since Distributed Training such as DDP internally
```

File: docs/source/common/test_set.rst [hunk 2]
```diff
@@ -144,5 +144,12 @@ Apart from this ``.validate`` has same API as ``.test``, but would rely respecti
     ``.validate`` method uses the same validation logic being used under validation happening within
     :meth:`~pytorch_lightning.trainer.trainer.Trainer.fit` call.
 
+.. warning::
+
+    When using ``trainer.validate()``, it is recommended to use a single device since Distributed Training such as DDP internally
```
Comment (carmocca) [inline docs/source/common/test_set.rst:149]:
```suggestion
    When using ``trainer.validate()``, it is recommended to use a single device since distributed strategies such as DDP internally
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 3]
```diff
@@ -382,14 +382,25 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
+            sampler = self._get_distributed_sampler(
                 dataloader,
                 shuffle,
                 mode=mode,
                 overfit_batches=self.trainer.overfit_batches,
                 **self.trainer.distributed_sampler_kwargs,
             )
 
+            trainer_fn = self.trainer.state.fn
+            if isinstance(sampler, DistributedSampler) and trainer_fn in (TrainerFn.VALIDATING, TrainerFn.TESTING):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:394]:
```suggestion
            if isinstance(sampler, DistributedSampler) and trainer.evaluating:
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 4]
```diff
@@ -382,14 +382,25 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
+                rank_zero_warn(
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:395]: Since this information is already printed here, do you still think it's worth adding the same to the docs?

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:395]: yeah, I think it is quite informative and good to have inside docs too for more visibility. If someone is just getting started and find this in docs, they might configure it accordingly then.

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 5]
```diff
@@ -382,14 +382,25 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
+                    f"Using `DistributedSampler` with the dataloaders. During trainer.{trainer_fn.value}(),"
+                    " it is recommended to use single device strategy to ensure each sample/batch gets evaluated"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:397]: Let's make it simple for the users

```suggestion
                    " it is recommended to use `devices=1` to ensure each sample/batch gets evaluated"
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 6]
```diff
@@ -382,14 +382,25 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
+                    " exactly once. Otherwise multi-device setting uses `DistributedSampler` replicates some samples"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:398]:
```suggestion
                    " exactly once. Otherwise, multi-device settings use `DistributedSampler` which replicates some samples"
```

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/data_connector.py:395]: If you prefer too have same message in both, then I'd suggest you add a comment here so we don't forget

File: docs/source/common/test_set.rst [hunk 7]
```diff
@@ -144,5 +144,12 @@ Apart from this ``.validate`` has same API as ``.test``, but would rely respecti
     ``.validate`` method uses the same validation logic being used under validation happening within
     :meth:`~pytorch_lightning.trainer.trainer.Trainer.fit` call.
 
+.. warning::
+
+    When using ``trainer.validate()``, it is recommended to use a single device since distributed strategies such as DDP
+    uses :class:`~torch.utils.data.distributed.DistributedSampler` internally, which replicates some samples to
+    make sure all devices have same batch size in case of uneven inputs. This is helpful to make sure
+    benchmarking for research papers is done the right way.
```
Comment (carmocca) [inline docs/source/common/test_set.rst:152]: Can you update this so it matches exactly the one in the code? (including my suggestions)

File: tests/trainer/connectors/test_data_connector.py [hunk 8]
```diff
@@ -66,3 +67,19 @@ def test_dataloader_source_request_from_module():
...
+    trainer = Trainer(strategy="ddp", devices=2, accelerator="cpu", fast_dev_run=True)
+    trainer._data_connector.attach_data(model)
+
+    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
+        trainer.state.fn = TrainerFn.VALIDATING
+        trainer.reset_val_dataloader(model)
+
+    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
+        trainer.state.fn = TrainerFn.TESTING
+        trainer.reset_test_dataloader(model)
```
Comment (carmocca) [inline tests/trainer/connectors/test_data_connector.py:85]: nit: cover only the minimum necessary

```suggestion
    trainer.state.fn = TrainerFn.VALIDATING
    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
        trainer.reset_val_dataloader(model)

    trainer.state.fn = TrainerFn.TESTING
    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
        trainer.reset_test_dataloader(model)
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 9]
```diff
@@ -382,14 +382,25 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
+                    " to make sure all devices have same batch size in case of uneven inputs."
+                )
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:400]: Should this be a "PossibleUserWarning"?

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:400]: yes! I always forgot that exists :p

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 10]
```diff
@@ -382,14 +382,27 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
                 mode=mode,
                 overfit_batches=self.trainer.overfit_batches,
                 **self.trainer.distributed_sampler_kwargs,
             )
 
+            # update docs too once this is resolved
+            trainer_fn = self.trainer.state.fn
+            if isinstance(sampler, DistributedSampler) and trainer_fn in (TrainerFn.VALIDATING, TrainerFn.TESTING):
+                rank_zero_warn(
+                    f"Using `DistributedSampler` with the dataloaders. During trainer.{trainer_fn.value}(),"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:397]:
```suggestion
                    f"Using `DistributedSampler` with the dataloaders. During `trainer.{trainer_fn.value}()`,"
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 11]
```diff
@@ -382,14 +382,27 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
                 shuffle,
                 mode=mode,
                 overfit_batches=self.trainer.overfit_batches,
                 **self.trainer.distributed_sampler_kwargs,
             )
 
+            # update docs too once this is resolved
+            trainer_fn = self.trainer.state.fn
+            if isinstance(sampler, DistributedSampler) and trainer_fn in (TrainerFn.VALIDATING, TrainerFn.TESTING):
+                rank_zero_warn(
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/data_connector.py:396]: I would make info. This is definitely a critical behavior for the users. This would be removed when we support uneven ddp.

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:396]: I would do only test for now and prevent DistributedSampler injection for testing.

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/data_connector.py:396]: I'd still consider using warning since this is critical, and an info might not reflect how important it is for benchmarking.

why not for `trainer.validate`? it uses the exact same API as that of `test`

**Review by justusschock — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: docs/source/common/test_set.rst [hunk 12]
```diff
@@ -50,8 +50,8 @@ To run the test set after training completes, use this method.
 
 .. warning::
 
-    It is recommended to test on single device since Distributed Training such as DDP internally
-    uses :class:`~torch.utils.data.distributed.DistributedSampler` which replicates some samples to
+    It is recommended to test with ``Trainer(devices=1)`` since distributed strategies such as DDP
+    uses :class:`~torch.utils.data.distributed.DistributedSampler` internally, which replicates some samples to
```
Comment (awaelchli) [inline docs/source/common/test_set.rst:54]:
```suggestion
    use :class:`~torch.utils.data.distributed.DistributedSampler` internally, which replicates some samples to
```
(plural)

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/data_connector.py:396]: +1 for warning

**Review by awaelchli — APPROVED** [review]
Like it :)

**Review by Borda — APPROVED** [review]
